src/Tools/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Codefix.Testing/BaseCodeFixTest.cs

```diff
@@ -0,0 +1,589 @@
+                return null;
+            }
+
+            bool done;
+            do
+            {
+                var analyzerDiagnostics = await GetSortedDiagnosticsFromDocumentsAsync(analyzers, project.Documents.ToArray(), cancellationToken).ConfigureAwait(false);
+                if (analyzerDiagnostics.Length == 0)
+                {
+                    break;
+                }
+
+                if (!AreDiagnosticsDifferent(analyzerDiagnostics, previousDiagnostics))
+                {
+                    break;
+                }
+
+                Verify.False(--numberOfIterations < -1, "The upper limit for the number of fix all iterations was exceeded");
+
+                Diagnostic firstDiagnostic = null;
+                string equivalenceKey = null;
+                foreach (var diagnostic in analyzerDiagnostics)
+                {
+                    if (!codeFixProvider.FixableDiagnosticIds.Contains(diagnostic.Id))
+                    {
+                        // do not pass unsupported diagnostics to a code fix provider
+                        continue;
+                    }
+
+                    var actions = new List<CodeAction>();
+                    var context = new CodeFixContext(project.GetDocument(diagnostic.Location.SourceTree), diagnostic, (a, d) => actions.Add(a), cancellationToken);
+                    await codeFixProvider.RegisterCodeFixesAsync(context).ConfigureAwait(false);
+                    if (actions.Count > (codeFixIndex ?? 0))
+                    {
+                        firstDiagnostic = diagnostic;
+                        equivalenceKey = actions[codeFixIndex ?? 0].EquivalenceKey;
+                        break;
+                    }
+                }
+
+                if (firstDiagnostic == null)
+                {
+                    numberOfIterations++;
+                    break;
+                }
+
+                previousDiagnostics = analyzerDiagnostics;
+
+                done = true;
+
+                FixAllContext.DiagnosticProvider fixAllDiagnosticProvider = TestDiagnosticProvider.Create(analyzerDiagnostics);
+
+                var analyzerDiagnosticIds = analyzers.SelectMany(x => x.SupportedDiagnostics).Select(x => x.Id);
+                var compilerDiagnosticIds = codeFixProvider.FixableDiagnosticIds.Where(x => x.StartsWith("CS", StringComparison.Ordinal));
+                var disabledDiagnosticIds = project.CompilationOptions.SpecificDiagnosticOptions.Where(x => x.Value == ReportDiagnostic.Suppress).Select(x => x.Key);
+                var relevantIds = analyzerDiagnosticIds.Concat(compilerDiagnosticIds).Except(disabledDiagnosticIds).Distinct();
+                var fixAllContext = new FixAllContext(project.GetDocument(firstDiagnostic.Location.SourceTree), codeFixProvider, scope, equivalenceKey, relevantIds, fixAllDiagnosticProvider, cancellationToken);
+
+                var action = await fixAllProvider.GetFixAsync(fixAllContext).ConfigureAwait(false);
```

Comment:
Note that this will fail for the Well-known [batch fixer](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/CodeFixes/FixAllOccurrences/WellKnownFixAllProviders.cs,24), which is used by large number of code fixers, if the test framework is targeting Microsoft.CodeAnalysis 2.6.0 or lower. This was fixed with https://github.com/dotnet/roslyn/pull/23492, and we recently moved roslyn-analyzers repo to 2.9 bits and now the batch fixer executes fine during our code fix tests.

Reply:
I would be fine implementing a workaround (MEF-exporting an implementation of the required interface from the test harness itself, as part of `CreateSolution`) or simply documenting it as a known issue.

Reply:
> simply documenting it as a known issue

We need to release a package based on a version of Microsoft.CodeAnalysis package that enables executing the well known batch fixer. That is the default batch fixer for our template and is used by majority of third party code fixers. Documenting it does not work is not useful as we are then not providing a solution to majority of our customers.

> StyleCop Analyzers bans the use of the well-known batch fixer outright due to its negative impact on test scenarios. Other analyzers that take a strict approach to compatibility would be likely to employ similar techniques, which this test harness helps support.

Approach taken by a specific third party analyzer package is completely irrelevant to the support being added here. We cannot add/block features from `Microsoft.CodeAnalysis.Testing` based on usage in StyleCop analyzers. If there are issues in the default Roslyn supported batch fixer, we either ought to fix it OR change the template to not advertise it. It makes no sense to have a code fix template that defaults to batch fixer and then release a test framework that explicitly does not support the batch fixer.

Reply:
> We need to release a package based on a version of Microsoft.CodeAnalysis package that enables executing the well known batch fixer.

If we can construct the MEF container in a manner that allows its use in testing older versions of Roslyn, it would certainly be an option. However, if it turns out this is not possible, it's also fine to ship a version that references older Roslyn APIs knowing that not all functionality is supported in those releases. As I mentioned previously, users who depend on newer Roslyn features with a minimum version greater than 1.0.1 will be able to apply assembly binding redirection without any negative impact on the tests. StyleCop Analyzers first introduced this strategy for churn testing after the 1.0.0 release and now depends on it as an integral part of its C# 7 test coverage.

Reply:
➡️ Moved to #138